cirq-core/cirq/ops/util.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, overload, TYPE_CHECKING, Union
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@overload
+def q(x: int) -> 'cirq.LineQubit':
+    ...
+
+
+@overload
+def q(x: int, y: int) -> 'cirq.GridQubit':
+    ...
+
+
+@overload
+def q(x: str) -> 'cirq.NamedQubit':
+    ...
+
+
+def q(
+    x: Union[int, str], y: Optional[int] = None
```

Comment:
Can we change the signature to accept `*args` and `**kwargs`, and still support mypy type checking using overloads, so that it's extensible in the future for a 4th qid type that we might introduce?

Reply:
I originally tried that but mypy wasn't happy with using `*args` with these overloads:
```
cirq-core/cirq/ops/util.py:36: error: Overloaded function implementation does not accept all possible arguments of signature 1  [misc]
cirq-core/cirq/ops/util.py:36: error: Overloaded function implementation does not accept all possible arguments of signature 2  [misc]
cirq-core/cirq/ops/util.py:36: error: Overloaded function implementation does not accept all possible arguments of signature 3  [misc]
```
I haven't been able to figure out a way around this. However, from my perspective adding additional signatures in the future, as long as we keep these signatures working and can make mypy happy, would be a backwards-compatible change.

One thing that's a bit unfortunate is I'd like to make these "positional-only" params so you can't do, say `q(x=1, y=0)`, but positional-only args are only supported in 3.8+. I will change the names to add an underscore prefix to indicate that these should be treated as positional-only.

Reply:
Success! The right incantation was to prefix arg names with _double_ underscore in the overload definitions and then mypy treats them as positional-only and allows using `*args`. Much cleaner this way, and more extensible for the future. Thanks for the suggestion!